# Louisville Free Public Library

Analysis of Young Adult (YA) genre in the Louisville Free Public Library collection.

## Questions

In this analysis we will look at the following questions:

- How much was spent on the collection for YA? 
- How many books are in the collection for YA?
- How does YA spending compare to other collections?
- Did the spending on YA change over time?
- Is YA more or less popular at any of the locations?

### Load the clean library collection data and show a preview of the data

In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

# load the clean books data into a dataframe and show the first few rows
books_data_path = Path('results/books_clean.csv.gz')
books_df = pd.read_csv(books_data_path)
books_df.head()

BibNum                                              Title  \
0  707409  Jeff Immelt and the new GE way : innovation, t...   
1  707411  Robin rescues dinner : 52 weeks of quick-fix m...   
2  707411  Robin rescues dinner : 52 weeks of quick-fix m...   
3  707411  Robin rescues dinner : 52 weeks of quick-fix m...   
4  707411  Robin rescues dinner : 52 weeks of quick-fix m...   

                 Author  PublicationYear ItemType     ItemCollection  \
0   Magee, David, 1965-             2009     Book  Adult Non-Fiction   
1  Miller, Robin, 1964-             2009     Book  Adult Non-Fiction   
2  Miller, Robin, 1964-             2009     Book  Adult Non-Fiction   
3  Miller, Robin, 1964-             2009     Book  Adult Non-Fiction   
4  Miller, Robin, 1964-             2009     Book  Adult Non-Fiction   

             ItemLocation  ItemPrice        Genre Audience  
0                    Main      25.95  Non-Fiction    Adult  
1               Southwest      19.99  Non-Fiction    Adult  
2               Southwest      19.99  Non-Fiction    Adult  
3  Remote Shelving - Main      19.99  Non-Fiction    Adult  
4  Remote Shelving - Main      19.99  Non-Fiction    Adult

### How much was spent on the collection for YA?

In [2]:
# TODO: First figure out which records in the dataframe are YA using a mask
# YA = Genre: Fiction, Audience = Teen. Then slice the DataFrame using the mask 
# and sum the ItemPrice column and format the result.
ya_mask = (books_df['Audience'] == 'Teen') & (books_df['Genre'] == 'Fiction')
ya_mask


0          False
1          False
2          False
3          False
4          False
           ...  
1187198    False
1187199    False
1187200    False
1187201    False
1187202    False
Length: 1187203, dtype: bool

In [5]:
"${:,}".format(books_df[ya_mask]['ItemPrice'].sum())

'$555,691.26'

The YA collection has a total cost of $555,691.26.

### How many books in  the collection are YA?

In [6]:
# TODO: Create a new column in the dataframe called YA_Category, use 
# value_counts() to get the count and percent of YA books, and use the concat()
# function to combibooks_df['YA_Category'] = np.where(ya_mask, 'YA', 'Other')
books_df['YA_Category'] = np.where(ya_mask, 'YA', 'Other')

ya_counts = books_df['YA_Category'].value_counts().apply(lambda x: "{:,}".format(x))
ya_percents = books_df['YA_Category'].value_counts(normalize=True).mul(100).round(1).astype(str) + '%'

pd.concat([ya_counts, ya_percents], axis=1, keys=['books','percentage'])

books percentage
Other  1,145,946      96.5%
YA        41,257       3.5%

YA accounts for 3.5% percent of the total number of books in the collection.

### How does YA spending compare to other collections?

In [7]:
# TODO: Group the data by Genre and Audience using groupby() and use sum() to 
# get the total cost. Format the totals as currency.
books_df.groupby(['Genre','Audience'])['ItemPrice'].sum().apply(lambda x: "${:,.2f}".format(x))

Genre        Audience
Fiction      Adult       $3,457,835.27
             Children      $687,553.59
             Teen          $555,691.26
             Unknown     $1,731,767.36
Non-Fiction  Adult       $9,209,529.31
             Children    $1,597,204.37
             Teen          $401,104.39
             Unknown       $875,794.32
Unknown      Adult         $281,617.43
             Children    $2,505,961.73
             Teen              $119.09
             Unknown       $533,619.31
Name: ItemPrice, dtype: object

### Did the spending on YA change over time?

In [26]:
# TODO: Calculate the counts (value_counts()), total cost (sum()), and average 
# cost (mean()) for all YA books by publication year using groupby(). 
# Concatenate the counts and costs using concat() into a single dataframe. 
# Format the counts and costs as numbers and currency.
ya_years_count = books_df[['PublicationYear', 'ItemPrice']][books_df['YA_Category']=='YA']\
.groupby('PublicationYear').count()
ya_years_count.columns = ['BookCount']

ya_years_total = books_df[['PublicationYear', 'ItemPrice']][books_df['YA_Category']=='YA']\
.groupby('PublicationYear').sum()
ya_years_total.columns = ['TotalCost']

ya_years_avg = books_df[['PublicationYear', 'ItemPrice']][books_df['YA_Category']=='YA']\
.groupby('PublicationYear').mean()
ya_years_avg.columns = ['AverageCost']

ya_years_summary = pd.concat([ya_years_count, ya_years_total, ya_years_avg], axis=1)

ya_years_summary['BookCount'] = ya_years_summary['BookCount'].apply(lambda x: "{:,}".format(x))
ya_years_summary['TotalCost'] = ya_years_summary['TotalCost'].apply(lambda x: "${:,}".format(x))
ya_years_summary['AverageCost'] = ya_years_summary['AverageCost'].apply(lambda x: "${:,}".format(x))

ya_years_summary

BookCount             TotalCost          AverageCost
PublicationYear                                                     
1919                    1                 $20.0                $20.0
1938                    3                $42.53  $14.176666666666668
1939                    5   $124.94999999999999               $24.99
1966                    1                 $2.99                $2.99
1967                    1                 $6.99                $6.99
1968                    2   $22.939999999999998  $11.469999999999999
1970                    3                $30.89  $10.296666666666667
1971                    8   $126.89999999999999  $15.862499999999999
1972                    2                $27.95              $13.975
1973                   15               $101.13               $6.742
1974                    6                $70.91  $11.818333333333333
1976                    7    $66.00999999999999   $9.429999999999998
1977                    9                $89.01                $9.89
1978                    8                $59.71             $7.46375
1979                    2                $20.98               $10.49
1980                    4                $60.58              $15.145
1981                    4                $40.92               $10.23
1982                    1                $13.49               $13.49
1983                    3                $11.57   $3.856666666666667
1984                    9   $57.870000000000005   $6.430000000000001
1985                   14               $141.15  $10.082142857142857
1986                    1                 $6.99                $6.99
1987                   15               $145.03   $9.668666666666667
1988                    6                $48.96                $8.16
1989                    8                $85.56              $10.695
1990                    8                $74.43             $9.30375
1991                   13               $131.82  $10.139999999999999
1992                   20               $150.04               $7.502
1993                   41               $336.98   $8.219024390243902
1994                   49               $575.11  $11.736938775510204
1995                   15               $155.69  $10.379333333333333
1996                   57                $576.4  $10.112280701754386
1997                   85             $1,122.03   $13.20035294117647
1998                   54               $577.92  $10.702222222222222
1999                  117             $1,392.21  $11.899230769230769
2000                  113             $1,404.27   $12.42716814159292
2001                  156             $1,824.02  $11.692435897435898
2002                  153             $1,868.17  $12.210261437908498
2003                  214             $2,636.94   $12.32214953271028
2004                  374             $4,228.69  $11.306657754010693
2005                  523             $6,984.28  $13.354263862332695
2006                  919            $11,682.18   $12.71183895538629
2007                1,185  $14,572.439999999999  $12.297417721518986
2008                1,304            $17,444.92  $13.378006134969324
2009                1,209            $16,679.89  $13.796435070306037
2010                1,577            $21,765.42  $13.801788205453391
2011                1,889  $28,039.019999999997  $14.843313922710427
2012                2,118  $31,727.659999999996  $14.980009442870632
2013                2,732  $40,770.549999999996  $14.923334553440702
2014                3,262            $50,246.38   $15.40354996934396
2015                3,371  $52,836.409999999996  $15.673808958765944
2016                2,916            $47,197.27  $16.185620713305898
2017                2,981             $46,134.2  $15.476081851727606
2018                3,483             $36,891.7  $10.591932242319839
2019                2,748            $28,427.17  $10.344676128093157
2020                2,790            $28,231.59  $10.118849462365592
2021          

Spending on YA books peaked in 2014 and 2015.

Is there a correlation between ItemPrice and PublicationYear?

In [29]:
# TODO: Use the corr() function to determine if there is a correlation between
# the ItemPrice and PublicationYear.

books_df[['ItemPrice','PublicationYear']].corr()

ItemPrice  PublicationYear
ItemPrice         1.000000        -0.256709
PublicationYear  -0.256709         1.000000

Price goes down over time, but the correlation is weak.

### Is YA more or less popular at any of the locations?

In [35]:
# TODO: Calculate the total number of books by location and the number of YA
# books by location and concatenate them into a single DataFrame. Add a new column
# to show the % of books by location that are YA. Format the values appropriately.

location_ya = books_df['ItemLocation'][books_df['YA_Category'] == 'YA'].value_counts()
location_ya.rename("YABookCount", inplace=True)

location_all = books_df['ItemLocation'].value_counts()
location_all.rename("TotalBookCount", inplace=True)

location_summary = pd.concat([location_all, location_ya], axis=1)

location_summary['PercentYA'] = (location_summary['YABookCount'] / location_summary['TotalBookCount'])

location_summary['TotalBookCount'] = location_summary['TotalBookCount']\
                                   .apply(lambda x: "{:,}".format(x, axis=1))
location_summary['YABookCount'] = location_summary['YABookCount']\
                                   .apply(lambda x: "{:,.0f}".format(x, axis=1))
location_summary['PercentYA'] = location_summary['PercentYA'].mul(100).round(1)
location_summary.sort_values(by=['PercentYA'], ascending=False)
location_summary

TotalBookCount YABookCount  PercentYA
Remote Shelving - Main           139,969          26        0.0
Northeast                        124,339       4,928        4.0
Southwest                        121,914       5,413        4.4
Main                             120,742          12        0.0
South Central                    115,614       5,238        4.5
Bon Air                           74,551       2,072        2.8
St Matthews                       69,417       2,636        3.8
Jeffersontown                     56,620       1,425        2.5
Iroquois                          52,190       1,910        3.7
Highlands - Shelby Park           45,352       1,457        3.2
Crescent Hill                     42,683       1,252        2.9
Childrens Main Library            38,668           1        0.0
Middletown                        33,069       1,333        4.0
Shively                           23,549       1,206        5.1
Newburg                           23,536       1,011        4.3
Fairdale                          23,025         983        4.3
Shawnee                           22,861       1,830        8.0
Western                           21,615       1,009        4.7
Portland                          13,313         491        3.7
Childrens Bookmobile               9,092         187        2.1
Remote Shelving - Shawnee          9,060       2,988       33.0
Main Teen                          6,018       3,848       63.9
Content Management                     4           1       25.0
Adult Bookmobile                       2         nan        NaN

Southwest is the location with the most YA books.